File: packages/asdl-pr-address/src/asdl_pr_address/cli/pr_address/prepare_run.py
"""Prepare a pr-address run by resolving PR context and normalizing feedback."""

from __future__ import annotations

import dataclasses
from pathlib import Path
from typing import Annotated, Any, Literal

import click
from pydantic import model_serializer

from asdl_core.clinkr.context import load_typed_context
from asdl_core.clinkr.exit import ClinkrExit
from asdl_core.clinkr.failure import ClinkrFailure
from asdl_core.clinkr.models import ClinkrModel
from asdl_core.clinkr.operation import clinkr_operation
from asdl_core.gh.types import (
    PRDiscussionComment,
    PRGatewayFailure,
    PRReview,
    PRReviewThread,
    PRState,
)
from asdl_core.git.types import DetachedHead, GitCommandFailure, RestructuredFile
from asdl_core.payloads.clinkr import open_clinkr_payload_store, write_clinkr_raw_payload_artifact
from asdl_core.payloads.store import PayloadStore
from asdl_pr_address.cli.pr_address.context import PrAddressCliContext
from asdl_pr_address.cli.pr_address.feedback_payload import (
    PayloadMode,
    PrepareRunPayloadManifest,
    build_prepare_run_payload_manifest,
)
from asdl_pr_address.cli.pr_address.prepare_run_workflow import (
    PreparedPrAddressRun,
    PrepareRunNoPr,
    PrepareRunPrLookupFailed,
    prepare_pr_address_run,
)

RestructuredFiles = tuple[RestructuredFile, ...]


def _gateway_failure_message(prefix: str, failure: PRGatewayFailure) -> str:
    detail = failure.stderr or failure.stdout or f"exit code {failure.returncode}"
    return f"{prefix}: {detail}"


class PrepareRunRequest(ClinkrModel):
    """Inputs to `prepare-run`.

    Attributes:
        include_all_threads: When True, include resolved reference threads in
            the normalized output. Default False returns only threads that
            still need classification (unresolved or reopened by this run).
        include_empty_reviews: When True, include PR-level reviews whose body
            is empty and whose state carries no request (`COMMENTED` /
            `APPROVED`). Default False drops them as noise; set True only when
            the caller wants to inspect the raw review stream.
    """

    include_all_threads: bool = False
    include_empty_reviews: bool = False
    payload_mode: Annotated[
        PayloadMode,
        click.Option(
            ["--payload-mode"],
            type=click.Choice(["inline", "payload"]),
            default="payload",
        ),
    ] = "payload"
    payload_session_id: Annotated[
        str | None,
        click.Option(["--payload-session-id"], type=click.STRING),
    ] = None


class PrepareRunInlineResult(ClinkrModel):
    """Normalized PR feedback snapshot for a single `pr-address` run.

    When `found` is False the current branch has no associated PR; all other
    fields except `current_branch`, `error`, and `returncode` are left at
    their defaults.

    Attributes:
        found: Whether a PR was resolved for the current branch.
        current_branch: The branch the helper was invoked on (None means
            detached HEAD, surfaced as a `ClinkrFailure` before this
            result is constructed).
        number: PR number on the host repository.
        title: PR title at the time of the snapshot.
        url: Web URL of the PR.
        head_ref_name: The PR's head branch — always equals `current_branch`
            for this flow.
        base_ref_name: The PR's base branch (e.g. "master"). Used by
            `get_restructured_files` to build the merge-base diff.
        state: PR lifecycle state (OPEN / CLOSED / MERGED).
        reviews: PR-level review submissions (the "Review changes" flow).
        review_threads: Normalized inline review threads. Contested threads
            previously resolved by `pr-address` are reopened and included
            here; resolved reference threads are included only when the
            request set `include_all_threads=True`.
        discussion_comments: Top-level PR discussion comments (not inline).
        reopened_thread_ids: Threads this run unresolved because a new
            comment landed after the pr-address resolution marker.
        restructured_files: Files renamed or copied between
            `origin/<base_ref_name>` and HEAD, detected via
            `git diff --name-status -M -C`. Used by the classifier to mark
            bot comments on moved code as `pre_existing`.
        warnings: Non-fatal issues encountered while preparing the run
            (e.g. `git diff` failed; failed to reopen a contested thread).
        error: stderr from `gh pr view` when `found` is False.
        returncode: exit code from `gh pr view` when `found` is False.
    """

    payload_mode: Literal["inline"] = "inline"
    found: bool
    current_branch: str | None = None
    number: int | None = None
    title: str | None = None
    url: str | None = None
    head_ref_name: str | None = None
    base_ref_name: str | None = None
    state: PRState | None = None
    reviews: tuple[PRReview, ...] = ()
    review_threads: tuple[PRReviewThread, ...] = ()
    discussion_comments: tuple[PRDiscussionComment, ...] = ()
    reopened_thread_ids: tuple[str, ...] = ()
    restructured_files: RestructuredFiles = ()
    warnings: tuple[str, ...] = ()
    error: str | None = None
    returncode: int | None = None

    @model_serializer
    def serialize_model(self) -> dict[str, Any]:
        if not self.found:
            payload: dict[str, Any] = {
                "payload_mode": self.payload_mode,
                "found": False,
                "current_branch": self.current_branch,
            }
            if self.error is not None:
                payload["error"] = self.error
            if self.returncode is not None:
                payload["returncode"] = self.returncode
            return payload

        return {
            "payload_mode": self.payload_mode,
            "found": True,
            "current_branch": self.current_branch,
            "number": self.number,
            "title": self.title,
            "url": self.url,
            "head_ref_name": self.head_ref_name,
            "base_ref_name": self.base_ref_name,
            "state": self.state,
            "reviews": [dataclasses.asdict(review) for review in self.reviews],
            "review_threads": [dataclasses.asdict(thread) for thread in self.review_threads],
            "discussion_comments": [
                dataclasses.asdict(comment) for comment in self.discussion_comments
            ],
            "reopened_thread_ids": list(self.reopened_thread_ids),
            "restructured_files": [dataclasses.asdict(item) for item in self.restructured_files],
            "warnings": list(self.warnings),
        }


@clinkr_operation(
    name="prepare-run",
    help="Resolve PR context, reopen contested threads, and normalize feedback.",
)
def run_prepare_run(
    ctx: click.Context,
    request: PrepareRunRequest,
) -> ClinkrExit[PrepareRunInlineResult | PrepareRunPayloadManifest]:
    store: PayloadStore | None = None
    if request.payload_mode == "payload":
        store = open_clinkr_payload_store(request.payload_session_id)

    pr_address_context = load_typed_context(ctx, PrAddressCliContext)
    outcome = prepare_pr_address_run(
        pr_address_context.pr_gateway,
        pr_address_context.git_gateway,
        cwd=Path.cwd(),
        include_all_threads=request.include_all_threads,
        include_empty_reviews=request.include_empty_reviews,
    )

    match outcome:
        case GitCommandFailure() as failure:
            raise ClinkrFailure(error_type="git_failed", message=failure.message)
        case DetachedHead():
            raise ClinkrFailure(
                error_type="detached_head",
                message="Detached HEAD: prepare-run requires a checked-out branch.",
            )
        case PrepareRunPrLookupFailed() as failure:
            raise ClinkrFailure(
                error_type="pr_gateway_failure",
                message=_gateway_failure_message(
                    f"Failed to look up PR for current branch {failure.current_branch!r}",
                    failure.failure,
                ),
            )
        case PrepareRunNoPr() as no_pr:
            inline_result = PrepareRunInlineResult(
                found=False,
                current_branch=no_pr.current_branch,
                error=no_pr.error,
                returncode=no_pr.returncode,
            )
        case PreparedPrAddressRun() as prepared:
            inline_result = PrepareRunInlineResult(
                found=True,
                current_branch=prepared.current_branch,
                number=prepared.number,
                title=prepared.title,
                url=prepared.url,
                head_ref_name=prepared.head_ref_name,
                base_ref_name=prepared.base_ref_name,
                state=prepared.state,
                reviews=prepared.reviews,
                review_threads=prepared.review_threads,
                discussion_comments=prepared.discussion_comments,
                reopened_thread_ids=prepared.reopened_thread_ids,
                restructured_files=prepared.restructured_files,
                warnings=prepared.warnings,
            )
    return _prepare_run_exit_for_payload_mode(
        inline_result=inline_result,
        payload_mode=request.payload_mode,
        store=store,
    )


def _prepare_run_exit_for_payload_mode(
    *,
    inline_result: PrepareRunInlineResult,
    payload_mode: PayloadMode,
    store: PayloadStore | None,
) -> ClinkrExit[PrepareRunInlineResult | PrepareRunPayloadManifest]:
    if payload_mode == "inline":
        return ClinkrExit.ok(inline_result)

    if store is None:
        raise AssertionError("payload artifact store must be opened before writing raw payload")

    descriptor = "pr-address-prepare-run-no-pr"
    if inline_result.found and inline_result.number is not None:
        descriptor = f"pr-address-prepare-run-pr-{inline_result.number}"
    raw_reference = write_clinkr_raw_payload_artifact(
        store=store,
        descriptor=descriptor,
        result=ClinkrExit.ok(inline_result),
    )
    return ClinkrExit.ok(
        build_prepare_run_payload_manifest(
            payload_reference=raw_reference,
            found=inline_result.found,
            current_branch=inline_result.current_branch,
            number=inline_result.number,
            title=inline_result.title,
            url=inline_result.url,
            head_ref_name=inline_result.head_ref_name,
            base_ref_name=inline_result.base_ref_name,
            state=inline_result.state,
            reviews=inline_result.reviews,
            review_threads=inline_result.review_threads,
            discussion_comments=inline_result.discussion_comments,
            reopened_thread_ids=inline_result.reopened_thread_ids,
            restructured_files=inline_result.restructured_files,
            warnings=inline_result.warnings,
            error=inline_result.error,
            returncode=inline_result.returncode,
        )
    )

File: ts/packages/pr-address/src/prepare-run.ts
import { z } from "zod";

import { failure, ok, toMachineEnvelope, type ClinkrExit, type ClinkrFailureExit } from "@asdl/clinkr";
import { defineExecOperation, gatewayFailureDetail, gatewayFailureMessage, gatewayOptions, type PrAddressExecContext } from "./exec-operation.ts";
import { contestedThreadIds, fetchFeedbackSnapshot } from "./feedback-collection.ts";
import type { GatewayFailure, PRDiscussionComment, PRReview, PRReviewThread, PRSummary, PrAddressGitGateway, PrAddressGitHubGateway, RestructuredFile } from "./gateways.ts";
import { buildPrepareRunPayloadManifest, type PayloadArtifactStore, type PayloadReference } from "./payload-store.ts";
import { compactOperationResult, stdoutModeSchema } from "./stdout-mode.ts";

interface PrepareRunInlineFound {
	payload_mode: "inline";
	found: true;
	current_branch: string;
	number: number;
	title: string;
	url: string;
	head_ref_name: string;
	base_ref_name: string;
	state: string;
	reviews: readonly PRReview[];
	review_threads: readonly PRReviewThread[];
	discussion_comments: readonly PRDiscussionComment[];
	reopened_thread_ids: readonly string[];
	restructured_files: readonly RestructuredFile[];
	warnings: readonly string[];
}

interface PrepareRunInlineNoPr {
	payload_mode: "inline";
	found: false;
	current_branch: string;
	error: string;
	returncode: number;
}

type PrepareRunInlineResult = PrepareRunInlineFound | PrepareRunInlineNoPr;

const prepareRunParseSchema = z.object({
	payload_mode: z.enum(["inline", "payload"]).default("payload"),
	harness_session_id: z.string().optional(),
	stdout_mode: stdoutModeSchema,
	include_all_threads: z.boolean().default(false),
	include_empty_reviews: z.boolean().default(false),
});

type PrepareRunRequest = z.output<typeof prepareRunParseSchema>;

export const prepareRunOperation = defineExecOperation({
	isRepoContextRequired: true,
	spec: {
		name: "prepare-run",
		description: "Resolve PR context, reopen contested threads, and normalize feedback.",
		schema: prepareRunParseSchema,
		handler: runPrepareRunOperation,
	},
});

async function runPrepareRunOperation(ctx: PrAddressExecContext, request: PrepareRunRequest): Promise<ClinkrExit<unknown>> {
	// Python opens the payload store before any gateway work; preserve that ordering.
	let store: PayloadArtifactStore | undefined;
	if (request.payload_mode === "payload" || request.stdout_mode === "compact") {
		const storeResult = await ctx.context.payloadStoreFactory.fromEnvironment({
			explicitHarnessSessionId: request.harness_session_id ?? null,
			env: ctx.env,
			clock: ctx.context.payloadClock,
		});
		if (storeResult.type === "error") return failure(storeResult.errorType, storeResult.message);
		store = storeResult.value;
	}

	const git = ctx.context.git;
	const branchResult = await git.getCurrentBranch(gatewayOptions(ctx));
	if (branchResult.type === "failure") return failure("git_failed", gitCommandFailureMessage(branchResult.failure));
	if (branchResult.type === "detached") return failure("detached_head", "Detached HEAD: prepare-run requires a checked-out branch.");
	const currentBranch = branchResult.branch;

	const github = ctx.context.github;
	const lookupResult = await github.getPrForBranch(currentBranch, gatewayOptions(ctx));
	if (lookupResult.type === "failure") {
		return failure("pr_gateway_failure", gatewayFailureMessage(`Failed to look up PR for current branch '${currentBranch}'`, lookupResult.failure));
	}

	let inlineResult: PrepareRunInlineResult;
	if (lookupResult.type === "miss") {
		inlineResult = { payload_mode: "inline", found: false, current_branch: currentBranch, error: lookupResult.stderr, returncode: lookupResult.returncode };
	} else {
		const prepared = await prepareFoundRun({
			ctx,
			pr: lookupResult.pr,
			currentBranch,
			github,
			git,
			shouldIncludeAllThreads: request.include_all_threads,
			shouldIncludeEmptyReviews: request.include_empty_reviews,
		});
		if (prepared.type === "error") return prepared.exit;
		inlineResult = prepared.value;
	}

	if (request.stdout_mode === "full") return ok(inlineResult);
	if (store === undefined) return ok(inlineResult);

	const descriptor = inlineResult.found ? `pr-address-prepare-run-pr-${inlineResult.number}` : "pr-address-prepare-run-no-pr";
	const rawReference = await store.writeJsonArtifact({
		descriptor,
		role: "raw",
		payload: toMachineEnvelope(ok(inlineResult)),
	});
	if (rawReference.type === "error") return failure(rawReference.errorType, rawReference.message);
	const manifest = buildManifest(inlineResult, rawReference.value);
	return ok(
		compactOperationResult({
			operation: "prepare-run",
			counts: inlineResult.found
				? {
						reviews: inlineResult.reviews.length,
						review_threads: inlineResult.review_threads.length,
						discussion_comments: inlineResult.discussion_comments.length,
						reopened_threads: inlineResult.reopened_thread_ids.length,
						restructured_files: inlineResult.restructured_files.length,
					}
				: { found: 0 },
			warnings: inlineResult.found ? inlineResult.warnings : [],
			artifacts: { full_output: rawReference.value },
			details: { manifest },
		}),
	);
}

async function prepareFoundRun(options: {
	ctx: PrAddressExecContext;
	pr: PRSummary;
	currentBranch: string;
	github: PrAddressGitHubGateway;
	git: PrAddressGitGateway;
	shouldIncludeAllThreads: boolean;
	shouldIncludeEmptyReviews: boolean;
}): Promise<{ type: "ok"; value: PrepareRunInlineFound } | { type: "error"; exit: ClinkrFailureExit }> {
	const snapshotResult = await fetchFeedbackSnapshot({
		gateway: options.github,
		prNumber: options.pr.number,
		shouldIncludeResolved: true,
		shouldIncludeEmptyReviews: options.shouldIncludeEmptyReviews,
		shouldCountAllReviewThreads: false,
		ctx: options.ctx,
	});
	if (snapshotResult.type === "error") return snapshotResult;
	const snapshot = snapshotResult.snapshot;

	const warnings: string[] = [];
	const reopenedThreadIds: string[] = [];
	for (const threadId of contestedThreadIds(snapshot.review_threads)) {
		const result = await options.github.unresolveReviewThread(threadId, gatewayOptions(options.ctx));
		if (result.type === "failure") {
			warnings.push(`Failed to reopen contested thread ${threadId}: ${gatewayFailureDetail(result.failure)}`);
			continue;
		}
		reopenedThreadIds.push(threadId);
	}

	const reopened = new Set(reopenedThreadIds);
	const normalizedThreads: PRReviewThread[] = [];
	for (const thread of snapshot.review_threads) {
		const adjustedThread = reopened.has(thread.id) ? { ...thread, is_resolved: false } : thread;
		if (options.shouldIncludeAllThreads || !adjustedThread.is_resolved) normalizedThreads.push(adjustedThread);
	}

	const filesResult = await options.git.getRestructuredFiles(options.pr.base_ref_name, gatewayOptions(options.ctx));
	let restructuredFiles: readonly RestructuredFile[];
	if (filesResult.type === "failure") {
		warnings.push(restructuredFilesFailureMessage(options.pr.base_ref_name, filesResult.failure));
		restructuredFiles = [];
	} else {
		restructuredFiles = filesResult.value;
	}

	return {
		type: "ok",
		value: {
			payload_mode: "inline",
			found: true,
			current_branch: options.currentBranch,
			number: options.pr.number,
			title: options.pr.title,
			url: options.pr.url,
			head_ref_name: options.pr.head_ref_name,
			base_ref_name: options.pr.base_ref_name,
			state: options.pr.state,
			reviews: snapshot.reviews,
			review_threads: normalizedThreads,
			discussion_comments: snapshot.discussion_comments,
			reopened_thread_ids: reopenedThreadIds,
			restructured_files: restructuredFiles,
			warnings,
		},
	};
}

function buildManifest(inlineResult: PrepareRunInlineResult, payloadReference: PayloadReference): unknown {
	if (!inlineResult.found) {
		return buildPrepareRunPayloadManifest({
			payload_reference: payloadReference,
			found: false,
			current_branch: inlineResult.current_branch,
			error: inlineResult.error,
			returncode: inlineResult.returncode,
		});
	}
	return buildPrepareRunPayloadManifest({
		payload_reference: payloadReference,
		found: true,
		current_branch: inlineResult.current_branch,
		number: inlineResult.number,
		title: inlineResult.title,
		url: inlineResult.url,
		head_ref_name: inlineResult.head_ref_name,
		base_ref_name: inlineResult.base_ref_name,
		state: inlineResult.state,
		reviews: inlineResult.reviews,
		review_threads: inlineResult.review_threads,
		discussion_comments: inlineResult.discussion_comments,
		reopened_thread_ids: inlineResult.reopened_thread_ids,
		restructured_files: inlineResult.restructured_files,
		warnings: inlineResult.warnings,
	});
}

/** Mirror the Python git gateway's current-branch failure message: stderr or a fixed fallback. */
function gitCommandFailureMessage(failure: GatewayFailure): string {
	return failure.stderr.trim() || "git failed";
}

/** Mirror the Python git gateway's restructured-files failure message for warning parity. */
function restructuredFilesFailureMessage(baseRefName: string, failure: GatewayFailure): string {
	return `Failed to detect restructured files against origin/${baseRefName}: ${failure.stderr.trim() || "git diff failed"}`;
}
